cmd/cloud-controller-manager/app/options/options.go

```diff
@@ -111,32 +107,36 @@ func NewCloudControllerManagerOptions() (*CloudControllerManagerOptions, error)
 }
 
 // NewDefaultComponentConfig returns cloud-controller manager configuration object.
-func NewDefaultComponentConfig(insecurePort int32) (componentconfig.CloudControllerManagerConfiguration, error) {
+func NewDefaultComponentConfig(insecurePort int32) (*componentconfig.CloudControllerManagerConfiguration, error) {
+	// TODO: This code will be fixed up/improved when the ccm API types are moved to their own, real API group out of
+	// pkg/apis/componentconfig to cmd/cloud-controller-manager/app/apis/
 	scheme := runtime.NewScheme()
 	if err := componentconfigv1alpha1.AddToScheme(scheme); err != nil {
-		return componentconfig.CloudControllerManagerConfiguration{}, err
+		return nil, err
 	}
 	if err := componentconfig.AddToScheme(scheme); err != nil {
-		return componentconfig.CloudControllerManagerConfiguration{}, err
+		return nil, err
 	}
+	scheme.AddKnownTypes(componentconfigv1alpha1.SchemeGroupVersion, &componentconfigv1alpha1.CloudControllerManagerConfiguration{})
+	scheme.AddKnownTypes(componentconfig.SchemeGroupVersion, &componentconfig.CloudControllerManagerConfiguration{})
 
-	versioned := componentconfigv1alpha1.CloudControllerManagerConfiguration{}
-	scheme.Default(&versioned)
-
-	internal := componentconfig.CloudControllerManagerConfiguration{}
-	if err := scheme.Convert(&versioned, &internal, nil); err != nil {
+	versioned := &componentconfigv1alpha1.CloudControllerManagerConfiguration{}
+	internal := &componentconfig.CloudControllerManagerConfiguration{}
+	scheme.Default(versioned)
+	if err := scheme.Convert(versioned, internal, nil); err != nil {
 		return internal, err
 	}
-	internal.KubeCloudShared.Port = insecurePort
+	internal.Generic.Port = insecurePort
 	return internal, nil
 }
 
 // Flags returns flags for a specific APIServer by section name
 func (o *CloudControllerManagerOptions) Flags() (fss apiserverflag.NamedFlagSets) {
-	o.CloudProvider.AddFlags(fss.FlagSet("cloud provider"))
-	o.Debugging.AddFlags(fss.FlagSet("debugging"))
-	o.GenericComponent.AddFlags(fss.FlagSet("generic"))
+	// TODO: Fix the --controllers flag for the ccm
+	o.Generic.AddFlags(fss.FlagSet("generic"), []string{}, []string{})
+	o.Generic.Debugging.AddFlags(fss.FlagSet("debugging"))
```

Comment:
This is strange that we call AddFlags on debugging. No sure how to factor that in a better way.

Reply:
this was done differently in https://github.com/kubernetes/kubernetes/pull/67469/commits/575d6d1d4b7ca63953197f5fe1726d717983a5dc#diff-adb0f59ace521b6de3bdb50a7d56c1d1R69, fyi

Reply:
edit: isn't this already done here: https://github.com/kubernetes/kubernetes/pull/66993/commits/e6bc13296a9ba7e13e6d47e5e739fe4a1cf507f2#diff-adb0f59ace521b6de3bdb50a7d56c1d1R69

Reply:
yes, this is wrong. We need another pattern to get the sectioned flags in a composable fashion. This is fine in a follow-up. But the linked line must be removed.